channels/layers.py

```diff
@@ -222,6 +238,85 @@ def __init__(self, *args, **kwargs):
             "".join(random.choice(string.ascii_letters) for i in range(12)),
         )
 
+    ### Expire cleanup ###
+
+    def _clean_expired(self):
+        """
+        Goes through all messages and groups and removes those that are expired.
+        Any channel with an expired message is removed from all groups.
+        """
+        with self.thread_lock:
+            # Channel cleanup
+            for channel, queue in list(self.channels.items()):
+                remove = False
+                # See if it's expired
+                while not queue.empty() and queue._queue[0][0] < time.time():
+                    queue.get_nowait()
+                    remove = True
+                # Any removal prompts group discard
+                if remove:
+                    self._remove_from_groups(channel)
+                # Is the channel now empty and needs deleting?
+                if not queue:
+                    del self.channels[channel]
```

Comment:
This method should also clean expired group memberships separately.

Reply:
What do you mean by that exactly?
Should I remove the channel from all groups if not queue?

Reply:
You need to go through the group memberships and expire the entries if their expiry time is bigger than `group_expiry` (there should be a test for this too, but I'll add that later). The Redis one does it here: https://github.com/django/channels_redis/blob/master/channels_redis/core.py#L316